Make `@fastmath` aware of `Base.literal_pow`

ajkeller34 opened the pull request:
as `Base.literal_pow(Base.FastMath.pow_fast, x, Val{p})`.

Thanks to PRs #20530 and #20889, exponentiation of a dimensionful quantity can be type-stable in some important circumstances. However, because macro expansion occurs before the `Val` lowering happens, `@fastmath x^2` cannot be type-stable for dimensionful quantities, whereas `x^2` could be. This is because `@fastmath x^2` is macro-expanded to `Base.FastMath.pow_fast(x,2)` and so we can never catch the `^` symbol. See the tests I added to get an idea of how this works.

Provided the premise of this PR isn't flawed somehow, it would probably good to have someone more familiar with `julia_syntax.scm` see if there’s a better way to implement this change. Because I tie into the mechanism introduced in #20889 I don't expect any "surprising behaviors" to be introduced here, though I haven't checked to see if there are any performance implications.

tkelman commented:
dunno if nanosoldier will really notice lowering performance, but may as well trigger @nanosoldier `runbenchmarks(ALL, vs = ":master")`

nanosoldier commented:
[Your benchmark job](https://github.com/JuliaLang/julia/pull/21099#issuecomment-287667646) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/3ce1a0597a7841a867feba0d9cc75e8ab5001b8f/5d65887_vs_268420d/report.md). cc @jrevels

TotalVerb commented:
It would make more sense to do this in the `@fastmath` macro than in lowering.

StefanKarpinski commented:
> It would make more sense to do this in the @fastmath macro than in lowering.

100% – we can't go around having special cases for every macro anyone might need special behavior for in the parser.

ajkeller34 commented:
Yeah, good point, I'll see what I can do.

ajkeller34 committed:
Make fastmath aware of Base.literal_pow.

ajkeller34 commented:
Updated, this approach seems much better and obvious in retrospect. I guess this is what happens when you learn about how lisp works for the first time and get a little too exuberant.

KristofferC commented:
Let's rerun CI and Nanosoldier and try get this merged.

KristofferC commented:
@nanosoldier `runbenchmarks(ALL, vs = ":master")`

stevengj commented:
It would be nice to have `@fastmath` convert `x^literal` to an optimal chain of multiplications ala #20637, but we can leave that for a future PR.

nanosoldier commented:
Something went wrong when running [your job](https://github.com/JuliaLang/julia/pull/21099#issuecomment-317033564):
```
NanosoldierError: failed to run benchmarks against primary commit: failed process: Process(`sudo cset shield -e su nanosoldier -- -c ./benchscript.sh`, ProcessExited(1)) [1]
```
Logs and partial data can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/3b37289200d4f00223a6ee8dc9f6b2eb6ef4108a/4e918b2_vs_ec5e6f7/report.md)
cc @ararslan

KristofferC commented:
Tests seems to fail now. Is it because the `Val` optimization is only made to "hardware numbertypes" now?

ajkeller34 commented:
I updated the PR, squashed, and rebased. `fastmath` tests passed locally. The problem was because the PR hadn't adopted the new `Val(x)` syntax from #22475.

ajkeller34 commented:
@KristofferC want to try triggering nanosoldier again? Looks like the tests passed (travis timed out for likely unrelated reasons).

ararslan commented:
Another Nanosoldier run is unlikely to work, still having problems

KristofferC commented:
I thought it worked a few times yesterday @nanosoldier `runbenchmarks(ALL, vs = ":master")`

nanosoldier commented:
[Your benchmark job](https://github.com/JuliaLang/julia/pull/21099#issuecomment-318808723) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/f8cbf80024c50a00824509af65ef590e839ef274/1fb1b03_vs_8d752a0/report.md). cc @ararslan

ararslan commented:
It's been working randomly but failing most of the time

ajkeller34 commented:
This is ready to merge as far as I can tell. Does anything else need to be done here?

musm commented on test/fastmath.jl:
```diff
@@ -202,3 +202,10 @@ let a = ones(2,2), b = ones(2,2)
 end
+
+struct LitPowTest end
+Base.literal_pow(::typeof(Base.FastMath.pow_fast), ::LitPowTest, ::Val{p}) where {p} = 1
```
does this mean if library authors want to use `^` with `@fastmath` they also have to add both
```
Base.literal_pow(::typeof(Base.FastMath.pow_fast),...
Base.literal_pow(::typeof(^),...
```

ajkeller34 replied:
@musm In most cases, no. There is no need for library authors to add any `literal_pow` methods unless they need to hook into the `literal_pow` mechanism.

Using this PR:
```
julia> struct Test end

julia> Base.:^(::Test, x::Int) = x

julia> t = Test()
Test()

julia> Base.FastMath.pow_fast(::Test, ::Any) = 5

julia> t^3
3

julia> @fastmath t^3
5
```

If you do hook into the `literal_pow` mechanism, then depending on what you're trying to achieve, you may or may not want to add both definitions to indicate what should be done with a literal int power when using `^` with or without  `@fastmath`. It is not required to define both. Continuing the previous example:

```
julia> Base.literal_pow(::typeof(Base.FastMath.pow_fast), ::Test, ::Val{p}) where {p} = 6

julia> @fastmath t^3.3
5

julia> @fastmath t^3
6

julia> x = 3
3

julia> @fastmath t^x
5

julia> t^3
3

julia> t^x
3
```

In this example, `@fastmath` does something special with a literal power, when regular exponentiation does not. You could of course define `Base.literal_pow(::typeof(^), ...)` to make exponentiation to a literal int pow without `@fastmath` behave as you wish.

tkelman replied:
that difference could perhaps use documentation somewhere

ajkeller34 replied:
It doesn't look like the `literal_pow` mechanism itself is documented, let alone how it works with `@fastmath`. Should I try to document all of that in this PR or in another?

tkelman replied:
another is fine. probably because the feature was considered experimental?

stevengj commented:
Bump.

stevengj commented:
Any chance of merging this?

JeffBezanson commented:
From triage: seems fine, just rebase please.

stevengj commented:
Rebased.

oscardssmith commented:
The test this PR adds is the one that's failing.

stevengj commented:
Interacting badly with #24240, will fix.

stevengj commented:
Actually, #24240 already seems to fix this issue in another way — `@fastmath x^2` goes to `pow_fast(x, Val{2}())` which calls `literal_pow` by default.

So maybe this can be closed?
